File: 2023/python/src/day1.py
import os
import sys
import re

def main():
    input = []
    
    with open(f'{os.path.dirname(os.path.realpath(__file__))}/../input/day1.txt') as f:
        for line in f.readlines():
            l = line.replace("\n","")
            input.append(l)

    print("Part 1: " + str(part1(input)))
    print("Part 2: " + str(part2(input)))

def part1(input):
    sum = 0
    for line in input:
        fst = findFirst(line)
        lst = findFirst(line[::-1])
        sum += int(fst+lst)
        
    return sum
      
def findFirst(str):
    for c in str:
        if c.isdigit():
            return c

def part2(input):
    sum = 0
    for line in input:
        fst = findFirst2(line)
        lst = findLast2(line)
        sum += int(fst+lst)
        
    return sum

def findFirst2(str):
    for i in range(0, len(str)):
        if str[i].isdigit():
            return str[i]
        else:
            for j in range(0, i):
                if toNum(str[j:i+1]):
                    return toNum(str[j:i+1])

def findLast2(str):
    for i in range(len(str)-1, -1, -1):
        if str[i].isdigit():
            return str[i]
        else:
            for j in range(len(str), i-1, -1):
                if toNum(str[i:j]):
                    return toNum(str[i:j])
        
def toNum(s):
    if s == "one":
        return "1"
    elif s == "two":
        return "2"
    elif s == "three":
        return "3"
    elif s == "four":
        return "4"
    elif s == "five":
        return "5"
    elif s == "six":
        return "6"
    elif s == "seven":
        return "7"
    elif s == "eight":
        return "8"
    elif s == "nine":
        return "9"
    else:
        return ""
    

if __name__ == "__main__":
    sys.exit(main())

File: 2023/haskell/day1.hs
import Data.Char
import Data.Maybe

part1 :: IO ()
part1 = do 
    contents <- readFile "input/day1.txt"
    let lines = words contents
    print . getSum $ lines

part2 :: IO ()
part2 = do
    contents <- readFile "input/day1.txt"
    let lines = words contents
    print . getSum2 $ lines

getSum :: [String] -> Int
getSum [xs] = read ([getFirst xs] ++ [getFirst (reverse xs)])
getSum (xs:xss) = read ([getFirst xs] ++ [getFirst (reverse xs)]) + getSum xss

getFirst :: String -> Char
getFirst [x] = x
getFirst (x:xs)
    | isDigit x = x
    | otherwise = getFirst xs

getSum2 :: [String] -> Int
getSum2 [xs] = read ([getFirst2 [] xs] ++ [getLast2 [] xs])
getSum2 (xs:xss) = read ([getFirst2 [] xs] ++ [getLast2 [] xs]) + getSum2 xss

getFirst2 :: String -> String -> Char
getFirst2 ys xs
    | isDigit . head $ xs = head xs
    | otherwise = case checkFirst [] (ys ++ [head xs]) of
        Just c -> c
        Nothing -> getFirst2 (ys ++ [head xs]) (tail xs)

getLast2 :: String -> String -> Char
getLast2 ys xs
    | isDigit . last $ xs = last xs
    | otherwise = case checkLast [] ([last xs] ++ ys) of
        Just c -> c
        Nothing -> getLast2 ([last xs] ++ ys) (init xs)

checkLast :: String -> String -> Maybe Char
checkLast ys [] = mapToNum ys
checkLast ys xs = case mapToNum ys of 
    Just c -> Just c
    Nothing -> checkLast (ys ++ [head xs]) (tail xs)

checkFirst :: String -> String -> Maybe Char
checkFirst ys [] = mapToNum ys
checkFirst ys xs = case mapToNum ys of 
    Just c -> Just c
    Nothing -> checkFirst ([last xs] ++ ys) (init xs)

mapToNum :: String -> Maybe Char
mapToNum xs
    | xs == "one" = Just '1'
    | xs == "two" = Just '2'
    | xs == "three" = Just '3'
    | xs == "four" = Just '4'
    | xs == "five" = Just '5'
    | xs == "six" = Just '6'
    | xs == "seven" = Just '7'
    | xs == "eight" = Just '8'
    | xs == "nine" = Just '9'
    | otherwise = Nothing